### PR TITLE
[AArch64][SME2] Enable bfm builtins for sme2

### PR DESCRIPTION
**Pull request by SamTebbs33**
This patch enables the following builtins for SME2
    svbfmlslb_f32
    svbfmlslb_lane_f32
    svbfmlslt_f32
    svbfmlslt_lane_f32

Patch by: Kerry McLaughlin <kerry.mclaughlin@arm.com>

### CODE REVIEW

File: clang/include/clang/Basic/arm_sve.td [hunk 1]
```diff
@@ -1992,3 +1986,12 @@ let TargetGuard = "sme2" in {
   def SVADD_SINGLE_X2 : SInst<"svadd[_single_{d}_x2]", "22d", "cUcsUsiUilUl", MergeNone, "aarch64_sve_add_single_x2", [IsStreaming], []>;
   def SVADD_SINGLE_X4 : SInst<"svadd[_single_{d}_x4]", "44d", "cUcsUsiUilUl", MergeNone, "aarch64_sve_add_single_x4", [IsStreaming], []>;
 }
+
+let TargetGuard = "sve2p1|sme" in {
```
Comment (MDevereau) [inline clang/include/clang/Basic/arm_sve.td:1990]:
```suggestion
let TargetGuard = "sve2p1|sme2" in {
```

Comment (SamTebbs33) [reply clang/include/clang/Basic/arm_sve.td]: Well spotted.

File: clang/test/CodeGen/aarch64-sve2p1-intrinsics/acle_sve2p1_bfmlsl.c [hunk 2]
```diff
@@ -2,10 +2,15 @@
 // REQUIRES: aarch64-registered-target
 
 // RUN: %clang_cc1 -fclang-abi-compat=latest -triple aarch64-none-linux-gnu -target-feature +sve2p1 -S -disable-O0-optnone -Werror -Wall -emit-llvm -o - %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s
+// RUN: %clang_cc1 -fclang-abi-compat=latest -triple aarch64-none-linux-gnu -target-feature +sme2 -target-feature +sve -S -disable-O0-optnone -Werror -Wall -emit-llvm -o - %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s
```
Comment (CarolineConcatto) [inline clang/test/CodeGen/aarch64-sve2p1-intrinsics/acle_sve2p1_bfmlsl.c:5]: Don't we need to add arm_streaming_somth here in order for these builtins to work with sme?

File: clang/include/clang/Basic/arm_sve.td [hunk 3]
```diff
@@ -1992,3 +1986,12 @@ let TargetGuard = "sme2" in {
   def SVADD_SINGLE_X2 : SInst<"svadd[_single_{d}_x2]", "22d", "cUcsUsiUilUl", MergeNone, "aarch64_sve_add_single_x2", [IsStreaming], []>;
   def SVADD_SINGLE_X4 : SInst<"svadd[_single_{d}_x4]", "44d", "cUcsUsiUilUl", MergeNone, "aarch64_sve_add_single_x4", [IsStreaming], []>;
 }
+
+let TargetGuard = "sve2p1|sme2" in {
+// == BFloat16 multiply-subtract ==
+  def SVBFMLSLB : SInst<"svbfmlslb[_{d}]", "dd$$", "f", MergeNone, "aarch64_sve_bfmlslb", [IsOverloadNone], []>;
```
Comment (CarolineConcatto) [inline clang/include/clang/Basic/arm_sve.td:1992]: Should we not add here that this is streaming compatible?
I can see: [CheckSVEEnabled](https://developer.arm.com/documentation/ddi0602/2022-09/Shared-Pseudocode/AArch64-Functions?lang=en#impl-aarch64.CheckSVEEnabled.0)();

File: clang/include/clang/Basic/arm_sve.td [hunk 4]
```diff
@@ -1992,3 +1986,12 @@ let TargetGuard = "sme2" in {
   def SVADD_SINGLE_X2 : SInst<"svadd[_single_{d}_x2]", "22d", "cUcsUsiUilUl", MergeNone, "aarch64_sve_add_single_x2", [IsStreaming], []>;
   def SVADD_SINGLE_X4 : SInst<"svadd[_single_{d}_x4]", "44d", "cUcsUsiUilUl", MergeNone, "aarch64_sve_add_single_x4", [IsStreaming], []>;
 }
+
+let TargetGuard = "sve2p1|sme2" in {
+// == BFloat16 multiply-subtract ==
+  def SVBFMLSLB : SInst<"svbfmlslb[_{d}]", "dd$$", "f", MergeNone, "aarch64_sve_bfmlslb", [IsOverloadNone, IsStreaming], []>;
+  def SVBFMLSLT : SInst<"svbfmlslt[_{d}]", "dd$$", "f", MergeNone, "aarch64_sve_bfmlslt", [IsOverloadNone, IsStreaming], []>;
+
+  def SVBFMLSLB_LANE : SInst<"svbfmlslb_lane[_{d}]", "dd$$i", "f", MergeNone, "aarch64_sve_bfmlslb_lane", [IsOverloadNone, IsStreaming], [ImmCheck<3, ImmCheck0_7>]>;
```
[the remark this answers is not in the document]
Comment (SamTebbs33) [reply clang/include/clang/Basic/arm_sve.td]: Thanks Sander. I've made these builtins use `IsStreamingCompatible`, which will be changed to `IsStreamingOrSVE2p1` once that is added.

**Comment by SamTebbs33** [comment]
Rebased to fix a merge conflict and made the builtins temporarily `IsStreamingCompatible`.

**Review by sdesmalen-arm — APPROVED** [review]
LGTM with comment addressed.

**Review by sdesmalen-arm — APPROVED** [review]

**Pull request merged**